File: src/client/controllers/CameraController/CameraPerspective.lua
local CameraPerspective = {};

local partsToMakeVisible = {"Left Arm", "Right Arm", "Left Leg", "Right Leg", "Torso"};

local function makeVisibleLocally(part: BasePart)
	if table.find(partsToMakeVisible, part.Name) then
		part.LocalTransparencyModifier = part.Transparency
		part.Changed:Connect(function(property)
			part.LocalTransparencyModifier = part.Transparency
		end)
	end
end

local function makeEquipmentVisibleLocally(equipment: Model)
	for _, part in equipment:GetChildren() do
		if (part:IsA("BasePart")) then
			part.LocalTransparencyModifier = part.Transparency
			part.Changed:Connect(function(property)
				part.LocalTransparencyModifier = part.Transparency
			end)
		end
	end
end

local function onCharacterAdded(character: Model)
	local humanoid = character:WaitForChild("Humanoid")
	for _,v in pairs(character:GetChildren()) do
		makeVisibleLocally(v)
	end
	
	for _, possibleEquipment in character:GetChildren() do
		if (possibleEquipment:HasTag("Equipment")) then
			makeEquipmentVisibleLocally(possibleEquipment)
		end
	end

	character.ChildAdded:Connect(function(child)
		if (child:HasTag("Equipment")) then
			makeEquipmentVisibleLocally(child)
		end
	end)
	
	humanoid.CameraOffset = Vector3.new(0, 0.1, -0.9);
	workspace.CurrentCamera.FieldOfView = 80
end

function CameraPerspective:init()
	local player = game:GetService("Players").LocalPlayer;
	
	if (player.Character) then
		onCharacterAdded(player.Character)
	end
	
	player.CharacterAdded:Connect(onCharacterAdded);
end

return CameraPerspective;

File: src/client/controllers/CameraController/CameraPerspective.ts
// CameraPerspective.ts

import { Workspace } from "@rbxts/services";

const partsToMakeVisible = ["Left Arm", "Right Arm", "Left Leg", "Right Leg", "Torso"];

function makeVisibleLocally(part: BasePart) {
	if (partsToMakeVisible.includes(part.Name)) {
		part.LocalTransparencyModifier = part.Transparency;
		(part.Changed as RBXScriptSignal<(property: string) => void>).Connect(() => {
            part.LocalTransparencyModifier = part.Transparency;
        });
	}
}

function makeEquipmentVisibleLocally(equipment: Model) {
	for (const part of equipment.GetChildren()) {
		if (part.IsA("BasePart")) {
			const basePart = part as BasePart;
			basePart.LocalTransparencyModifier = basePart.Transparency;
			(basePart.Changed as RBXScriptSignal<(property: string) => void>).Connect(() => {
				basePart.LocalTransparencyModifier = basePart.Transparency;
			});
		}
	}
}

function onCharacterAdded(character: Model) {
	const humanoid = character.WaitForChild("Humanoid") as Humanoid;
	for (const v of character.GetChildren()) {
		if (v.IsA("BasePart")) {
			makeVisibleLocally(v as BasePart);
		}
	}

	for (const possibleEquipment of character.GetChildren()) {
		if ((possibleEquipment as Instance & { HasTag?: (tag: string) => boolean }).HasTag?.("Equipment")) {
			makeEquipmentVisibleLocally(possibleEquipment as Model);
		}
	}

	character.ChildAdded.Connect((child) => {
		if ((child as Instance & { HasTag?: (tag: string) => boolean }).HasTag?.("Equipment")) {
			makeEquipmentVisibleLocally(child as Model);
		}
	});

	humanoid.CameraOffset = new Vector3(0, 0.1, -0.9);
	(Workspace.CurrentCamera as Camera).FieldOfView = 80;
}

const CameraPerspective = {
	init() {
		const player = game.GetService("Players").LocalPlayer;
		if (player.Character) {
			onCharacterAdded(player.Character);
		}
		player.CharacterAdded.Connect(onCharacterAdded);
	}
};

export = CameraPerspective;